### ISSUE TITLE
No cbm_k_tksa

### ISSUE THREAD

**Issue by dmantione** [cc65/cc65#525]
Hello,

I am missing another essential KERNAL call in cbm.h: There is no cbm_k_tksa(). This time it seems to be really missing, there is no trace of it in the source. The call is essential because after a TALK, the next step to do in serial bus communication is to transmit a secondary address on the serial bus, i.e. cbm_k_talk() is useless without it.

For example, a routine to read the error channel 15 from a floppy drive is as follows:
```
char drive_error_code;

char device_read_error(char device) {
  char* p;
  cbm_k_talk(device);
  asm ("lda #$6F"); /* $60 + $0F commands secondary channel $0F to talk to the computer */
  asm ("jsr $FF96");
  p=(char *)0x200;
  do
    *p=cbm_k_acptr();
  while (*p++!=13);
  *p=0;
  cbm_untlk(device);
//  printf("%s\n",0x200);
  p=(char *)0x200;
  drive_error_code=10*(p[0]-0x30)+(p[1]-0x30);
  return drive_error_code;
}
```
The use of inline assembler is ugly, so it would be usefull to have a cbm_k_tksa().

**Comment by spiro-trikaliotis** [comment]
You're right.
And cbm_k_second (for the secondary address after listen) is also missing.
And cbm_k_untlk
And some others, too.

It seems one should go through the list of KERNAL routines and add the ones that are missing.

**Comment by oliverschmidt** [comment]
Contributions welcome...

**Comment by Compyx** [comment]
This looks like something I might be able to help with, doesn't appear to be too complicated. I already implemented the cbm_k_tska() function once to get a feel for the cc65 codebase.

But before I start, any pointers to documentation on how the cc65 codebase is structured, code style guidelines, that sort of stuff?

**Comment by oliverschmidt** [comment]
For of all thanks for your interest in contributing to cc65 :-)

If you check out https://github.com/cc65/cc65/tree/master/libsrc/cbm there are a bunch of files named c_*.s - I'd say whatever you do should fit in there seamlessly and that's all...

**Comment by Compyx** [comment]
Looks like adding these missing Kernal calls will mostly involve copy-pasting from already existing files. And of course some proper testing.

I'll have a go at it :)

**Comment by greg-king5** [comment]
The style guides are in the wiki:
https://github.com/cc65/wiki/wiki#this-cc65-projects-coding-guidelines

**Comment by Compyx** [comment]
Thanks for the Wiki link, that should be enough to get me started.

**Comment by Compyx** [comment]
I do indeed intend to create a pull request once I have some code ready. Unfortunately I got a little sidetracked with getting the new VICE release ready :)

**Comment by Compyx** [comment]
Going through the list of kernal calls of the C64, using my C64 Programmer's Reference Guide, it appears some kernal calls are misnamed. For example CHKOUT (PRG) is called CKOUT in libsrc/c64/kernal.s and in the C library is called cbm_k_ckout() (in libsrc/cbm/c_ckout.c). Same goes for CLRCHN (PRG) vs CLRCH (cc65).

So how would one handle this? Renaming the functions and labels would appear to be the obvious choice, but would break any code using the 'old' names.

Is there any definitive list of how CBM kernal calls are supposed to be called? I tend to use the C64 PRG as my reference, but that's because the C64 is the only CBM machine I have actual (non-emu) experience with.

So before I go on adding more kernal calls to cbm.h, I'd like to know how to proceed with misnamed kernal calls in the cc65 codebase.

**Comment by polluks** [comment]
@Compyx clrch and clrchn both exist http://www.zimmers.net/cbmpics/cbm/PETx/petmem.txt

**Comment by oliverschmidt** [comment]
As always I'm personally not too much into backward compatibility. Most users stay with obsolete cc65 versions anyway, totally independent how much we care for them - or ignore them. And the people actually "following" the cc65 development expect incompatible changes here and there.

So if - and only if - the SMEs (knew it? It's "subject matter experts" ;-) agree that something is named wrong I personally opt to correct it. However, I'll for sure (at least try to) keep my hands out of CBM nomenclature discussions - I already burned them ;-)

**Comment by dmantione** [comment]
I would say that backward compatibility is no big deal, because calls that are essential for programs were missing: There cannot be a huge amount of code depending on this, otherwise this would have been flagged before.

But I just the C64 programmers reference guide is too little evidence that the naming is incorrect. I haven't found any evidence for "ckout", but look for example at Microsofts Basic source code:

https://github.com/mist64/c64rom/blob/master/c64rom_ms.txt

... Microsoft did call this cq0out! So there exist multiple naming conventions. A 5 character identifier limit in an ancient assembler would be a plausible explanation for "ckout".

**Comment by Compyx** [comment]
Aliasing a function could be as simple as:
```
#ifndef cbm_k_chrout
#define cbm_k_chrout(C) cbm_k_bsout(C)
#endif
```

**Comment by oliverschmidt** [comment]
I personally have no problem with as many aliases as the SMEs see fit...

**Comment by silverdr** [comment]
I /think/ there are mainly two (official) schemes. One stemming from the earlier PET times, probably limited to five characters due to old assembler limitations as @dmantione  mentioned. And the newer one, which can be found in official publications for the 64 ("[PRG](http://e4aws.silverdr.com/project64/c64prg/)", "[Mapping](http://e4aws.silverdr.com/project64/mapc64/)"), which IMHO should be available *at least* for the 64 and newer machines. Obviously aliasing them for all CBMs isn't going to hurt either.

**Comment by Compyx** [comment]
When I do add aliasing, it might be a good idea to update the documentation to explain this, and providing a proper list of kernal calls supported by each CBM target. Right now the documentation on "CBM-specific functions" mentions "Some functions are available for all (or at least most) of the Commodore machines.". It might be better to provide an exact list of which kernal calls are available for each CBM target.

Then there's the 'overlap' in the various kernal jump tables, for example: on the C64 $FF96 is TKSA (cbm_k_tksa) but on the PET $FF96 is the BASIC v4 DOPEN call (not implemented yet afaik). The linker "solves" this issue by not exporting TKSA when using the 'cbm2' target, but again, documenting this couldn't hurt. (And yes, I'm pedantic when it comes to documentation :))

Let's see what the Subject Matter Experts have to say, there seem to be plenty around: I noticed quite a few VICE contributors also contributing here.

**Comment by silverdr** [comment]
I imagine this being documented somewhat like

* NEWERNAME (OLDERNAME) ADDRESS0_if_applicable TARGETs - description ...

If we do it by names, then address conflicts won't matter since "TARGETs" will inform which targets particular name at particular address pertain to. If we have the same name at different addresses for different targets - that will also be covered. I don't remember how I put the KERNAL calls descriptions into the docs (AFAIR I put most of those in some good time ago) but it should be adaptable.

**Comment by Compyx** [comment]
I managed to add the documentation for the calls I added just fine, looking at what was already there. I've never worked with linuxdoc-tools before, but I found it pretty easy to get into. And since `make doc` takes little time, it's easy to locate any errors and fix them.

**Comment by oliverschmidt** [comment]
> And since make doc takes little time, it's easy to locate any errors and fix them.

:-)

**Comment by greg-king5** [comment]
Fixed by PRs #656 and #662.  Closing this issue.
